File: patternlets/MPI/08.broadcast2/broadcast2.c
/* broadcast2.c
 * ... illustrates the use of MPI_Bcast() for arrays...
 * Joel Adams, Calvin College, November 2009.
 * Steven R. Vegdahl, University of Portland, 2016
 *   (update to print function)
 *
 * Usage: mpirun -np N ./broadcast2
 *
 * Exercise:
 * - Compile and run, using 2, 4, and 8 processes 
 * - Use source code to trace execution and output
 * - Explain behavior/effect of MPI_Bcast().
 */

#include <mpi.h>        // MPI functions
#include <stdio.h>      // printf, sprintf, ...
#include <limits.h>     // CHAR_BIT

#define MAX 8

void fill(int* a, int size);
void print(char* str, int id, int* a, int size);


int main(int argc, char** argv) {
  int array[MAX] = {0};
  int numProcs, myRank;
  
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &numProcs);
  MPI_Comm_rank(MPI_COMM_WORLD, &myRank);
  
  if (myRank == 0) fill(array, MAX);
  
  print("BEFORE", myRank, array, MAX);
  
  MPI_Bcast(array, MAX, MPI_INT, 0, MPI_COMM_WORLD);
  
  print("AFTER", myRank, array, MAX);
  
  MPI_Finalize();
  
  return 0;
}


/* fill an array with some values
 * @param: a, an int*.
 * @param: size, an int.
 *
 * Precondition: a is the address of an array of ints.
 *  && size is the number of ints a can hold.
 *
 * Postcondition: a has been filled with values
 * { 11, 12, 13, ... }.
 */
void fill(int* a, int size) {
  int i;
  for (i = 0; i < size; i++) {
    a[i] = i+11;
  }
}


/* display a string, a process id, and its array values
 * @param: str, a char*
 * @param: id, an int 
 * @param: a, an int*
 * @param: size, an int.
 *
 * Precondition: str points to either "BEFORE" or "AFTER"
 *  && id is the rank of this MPI process
 *  && a is the address of a MAX-element int array.
 *
 * Postcondition: str, id, and a have all been written to stdout.
 */

#define MAX_INT_PRINT_SIZE ((CHAR_BIT*sizeof(int))+1)

void print(char* str, int id, int* a, int size) {

  // sprint-buffer, including space for separators and label-strings
  // assumes that 'str' length is small (e.g., < 10)
  char buffer[(size+1)*(MAX_INT_PRINT_SIZE+2)+100];

  // internal buffer pointer 
  char *spot = buffer;
  
  // sprint the characters before the elements, updating position
  spot += sprintf(spot, "%s broadcast, process %d has: {", str, id);
  
  // separator string--set to empty string to avoid putting one
  // before the first element
  char *sep = "";
  
  // sprint each element, bumping internal buffer pointer
  for (int i = 0; i < size; i++) {
    spot += sprintf(spot, "%s%d", sep, a[i]);
    sep = ", ";
  }
  
  // sprint the stuff after the elements
  spot += sprintf(spot, "}\n");
  
  // actually print out the string to stdout
  printf("%s", buffer);
}

File: patternlets/java-OpenMPI/12.broadcast2/Broadcast2.java
/* Broadcast.java
 * ... illustrates the use of MPI's broadcast command with multiple values.
 *
 * Note: This version uses an IntBuffer to store the values.
 *
 * Goal: The master process reads an 'answer' value from a file
 *        and broadcasts it to all the other processes.
 *       Each process outputs its 'answer' value before and after
 *        the broadcast.
 *
 * Joel Adams, Calvin University, November 2019,
 *
 * Usage: mpirun -np 4 java ./Broadcast2
 *
 * Exercise:
 * - Compile, then run several times,
 *     using 2, 4, and 8 processes
 * - Use source code to trace execution and output
 * - Explain behavior/effect of MPI_Bcast().
 */

import java.nio.IntBuffer;
import mpi.*;

public class Broadcast2 {

 public static void main(String [] args) throws MPIException {
    MPI.Init(args);

    Comm comm         = MPI.COMM_WORLD;
    int id            = comm.getRank();

    IntBuffer buffer = MPI.newIntBuffer(BUFFER_SIZE);  // all: allocate buffers

    if ( id == MASTER ) {                              // MASTER: fill its buffer
        fill(buffer);
    }

    print("BEFORE", id, buffer);                       // all: print buffers before

    printSeparator("----", id, comm);

    comm.bcast(buffer, buffer.capacity(), MPI.INT, 0); // all: participate in broadcast

    print("AFTER", id, buffer);                        // all: print buffers after

    MPI.Finalize();
  }


  /* utility to fill an array with some values.
   * @param: a, an int array.
   * POST: a has been filled with int values.
   */
  private static void fill(IntBuffer buf) {
     for (int i = 0; i < buf.capacity(); ++i) {
         buf.put(i, i + 11);
     }
  }

  /* utility to print a buffer with descriptive labels.
   * @param: label, a String.
   * @param: id, this process's MPI rank.
   * @param, buf, an IntBuffer.
   * POST: label, id, and buf have been displayed via System.out.
   */
  private static void print(String label, int id, IntBuffer buf) {
    String msg = label + " the broadcast, process " + id
                       + "'s buffer contains:";
    for (int i = 0; i < buf.capacity(); ++i) {
      msg += (" " + buf.get(i));
    }
    msg += "\n";
    System.out.print(msg); 
  }
 
  /* utility to print a separator string between the 'before' and 'after' parts.
   * @param: separator, a String.
   * @param: id, the rank of this MPI process.
   * @param: comm, the Communicator for the processes involved.
   * POST: the master has printed the separator to System.out.
   */
  public static void printSeparator(String separator, int id, Comm comm) 
                                      throws MPIException {
     comm.barrier();
     if (id == MASTER) { System.out.println(separator); }
     comm.barrier();
  }

  private static final int MASTER = 0;
  private static final int BUFFER_SIZE = 8;
}

